internal/storage/storage.go

```diff
@@ -39,72 +66,182 @@ func NewDataWriter(filename string) (*DataWriter, error) {
 	}, nil
 }
 
-func (p *DataWriter) WriteData(record *Record) error {
-	return p.encoder.Encode(record)
+func (w *DataWriter) WriteData(record *Record) error {
+	return w.encoder.Encode(record)
 }
 
-func (p *DataWriter) Close() error {
-	return p.file.Close()
+func (w *DataWriter) Close() error {
+	return w.file.Close()
 }
 
-func NewStorage(filename string) (*Storage, error) {
+func NewStorage(filename, connectionData string) (*Storage, error) {
 	var dataWriter *DataWriter
+	var db *Database
+	var err error
+
+	mode := MemoryMode
+
+	if connectionData != "" {
+		mode = DBMode
+		db, err = NewDatabase(connectionData)
+		if err != nil {
+			return nil, err
+		}
+	} else if filename != "" {
+		mode = FileMode
+	}
 
 	store := Storage{
 		Links:      make(map[string]string),
 		filename:   filename,
 		dataWriter: dataWriter,
+		database:   db,
+		mode:       mode,
 	}
 
 	return &store, nil
 }
 
 func (s *Storage) Restore() error {
-	file, err := os.OpenFile(s.filename, os.O_RDONLY|os.O_CREATE, 0666)
-	if err != nil {
-		return err
-	}
-	defer file.Close()
+	switch s.mode {
+	case FileMode:
+		file, err := os.OpenFile(s.filename, os.O_RDONLY|os.O_CREATE, 0666)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
 
-	fileScanner := bufio.NewScanner(file)
+		fileScanner := bufio.NewScanner(file)
 
-	for fileScanner.Scan() {
-		var record Record
-		line := fileScanner.Text()
-		err = json.Unmarshal([]byte(line), &record)
+		for fileScanner.Scan() {
+			var record Record
+			line := fileScanner.Text()
+			err = json.Unmarshal([]byte(line), &record)
+			if err != nil {
+				logger.Log.Error("data decoding problem", zap.Error(err))
+				continue
+			}
+
+			s.Links[record.ShortULR] = record.OriginalURL
+		}
+	case DBMode:
+		err := s.database.CreateDatabase()
 		if err != nil {
-			logger.Log.Error("data decoding problem", zap.Error(err))
-			continue
+			logger.Log.Error("database creation error", zap.Error(err))
+			return err
 		}
-
-		s.Links[record.ShortULR] = record.OriginalURL
 	}
 
 	return nil
 }
 
-func (s *Storage) Add(key, value string) {
-	id := uuid.New()
+func (s *Storage) Add(key, value string) error {
+	id := uuid.NewString()
 	record := Record{
 		UUID:        id,
 		ShortULR:    key,
 		OriginalURL: value,
 	}
 
-	if s.filename != "" {
+	if s.mode == DBMode {
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		defer cancel()
+		err := s.database.SaveRecord(ctx, &record)
+		if err != nil {
+			var pgErr *pgconn.PgError
+			if errors.As(err, &pgErr) && pgerrcode.IsIntegrityConstraintViolation(pgErr.Code) {
+				return ErrConflict
+			} else {
+				logger.Log.Error("save data to database error", zap.Error(err))
+				return err
+			}
```

Comment:
```suggestion
			} 
			logger.Log.Error("save data to database error", zap.Error(err))
			return err
			
```
лишние else
в целом желательно всегда стараться обходиться с минимальным количеством условий